nilearn/surface/surface.py

```diff
@@ -1975,3 +1976,243 @@ def new_img_like(ref_img, data):
         mesh=copy.deepcopy(mesh),
         data=data,
     )
+
+
+def _compute_adjacency_matrix(mesh, values="ones", dtype=None):
+    """Compute the adjacency matrix for a surface.
+
+    The adjacency matrix is a matrix
+    with one row and one column for each vertex
+    such that the value of a cell `(u,v)` in the matrix is 1
+    if nodes `u` and `v` are adjacent and 0 otherwise.
+
+    Parameters
+    ----------
+    mesh : InMemoryMesh
+
+    values : { 'len' | 'invlen' | 'ones'}, optional
+        If `values` is `'ones'` (the default), then the returned matrix
+        contains uniform values in the cells representing edges.
+        If the value is `'len'` then the cells contain
+        the edge length of the represented edge.
+        If the value is `'invlen'`, then the the inverse of the distances
+        are returned.
+
+    dtype : numpy dtype-like or None, optional
+        The dtype that should be used for the returned sparse matrix.
+
+    Returns
+    -------
+    matrix : scipy.sparse.csr_matrix
+        A sparse matrix representing the edge relationships in `surface`.
+
+    """
+    # This is a bit of a hack to quickly find a unique set of all edges.
+    if values not in {"len", "invlen", "ones"}:
+        raise ValueError(f"unrecognized values argument: {values}")
+
+    n = mesh.coordinates.shape[0]
+
+    edges = np.vstack(
+        [
+            mesh.faces[:, [0, 1]],
+            mesh.faces[:, [0, 2]],
+            mesh.faces[:, [1, 2]],
+        ]
+    )
+    edges = edges.astype(np.int64)
+    bigcol = edges[:, 0] > edges[:, 1]
+    lilcol = ~bigcol
+    edges = np.concatenate(
+        [
+            edges[bigcol, 0] + edges[bigcol, 1] * n,
+            edges[lilcol, 1] + edges[lilcol, 0] * n,
+        ]
+    )
+    edges = np.unique(edges)
+
+    (u, v) = (edges // n, edges % n)
+
+    # Calculate distances between pairs.
+    # We use this as a weighting to make sure that
+    # smoothing takes into account the distance between each vertex neighbor
+    if values in ("len", "invlen"):
+        coords = mesh.coordinates
+        edge_lens = np.sqrt(np.sum((coords[u, :] - coords[v, :]) ** 2, axis=1))
+        if dtype is None:
+            dtype = edge_lens.dtype
+        else:
+            edge_lens = edge_lens.astype(dtype)
+        if values == "invlen":
+            edge_lens = 1 / edge_lens
+    elif values == "ones":
+        if dtype is None:
+            edge_lens = np.ones_like(edges)
+        else:
+            edge_lens = np.ones(edges.shape, dtype=dtype)
+
+    # We can now make a sparse matrix.
+    ee = np.concatenate([edge_lens, edge_lens])
+    uv = np.concatenate([u, v])
+    vu = np.concatenate([v, u])
+
+    return csr_matrix((ee, (uv, vu)), shape=(n, n))
+
+
+def smooth_img(
+    imgs,
+    iterations=1,
+    distance_weights=False,
+    vertex_weights=None,
+    center_surround_knob=0,
+    match="sum",
```

Comment:
if we decide to keep the rescaling then maybe let's use another word than "match" that is now a [soft keyword for python >=  3.10](https://peps.python.org/pep-0622/)